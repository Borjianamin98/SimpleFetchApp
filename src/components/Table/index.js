import React from "react";
import {SORTS} from "../../constants";
import Button from "../Button";
import PropTypes from "prop-types";
import classNames from "classnames";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faSort, faSortDown, faSortUp} from "@fortawesome/free-solid-svg-icons";

const largeColumn = {
    width: '40%',
};
const midColumn = {
    width: '30%',
};
const smallColumn = {
    width: '10%',
};

const Table = ({list, onDismiss}) => {
    const [sortKey, setSortKey] = React.useState('NONE');
    const [isSortReverse, setIsSortReverse] = React.useState(false);

    function onSort(newSortKey) {
        setIsSortReverse(sortKey === newSortKey && !isSortReverse)
        setSortKey(newSortKey);
    }

    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse
        ? sortedList.reverse()
        : sortedList;

    return <div className="table">
        <div className="table-header">
        <span style={{width: '40%'}}>
            <Sort
                sortKey={'TITLE'}
                onSort={onSort}
                activeSortKey={sortKey}
                isSortReverse={isSortReverse}
            >
            Title
            </Sort>
        </span>
            <span style={{width: '30%'}}>
            <Sort
                sortKey={'AUTHOR'}
                onSort={onSort}
                activeSortKey={sortKey}
                isSortReverse={isSortReverse}
            >
            Author
            </Sort>
        </span>
            <span style={{width: '10%'}}>
            <Sort
                sortKey={'COMMENTS'}
                onSort={onSort}
                activeSortKey={sortKey}
                isSortReverse={isSortReverse}
            >
            Comments
            </Sort>
        </span>
            <span style={{width: '10%'}}>
            <Sort
                sortKey={'POINTS'}
                onSort={onSort}
                activeSortKey={sortKey}
                isSortReverse={isSortReverse}
            >
            Points
            </Sort>
        </span>
            <span style={{width: '10%'}}>
            Archive
        </span>
        </div>
        {reverseSortedList.map(item =>
            <div key={item.objectID} className="table-row">
            <span style={largeColumn}>
                <a href={item.url}>{item.title}</a>
            </span>
                <span style={midColumn}>{item.author}</span>
                <span style={smallColumn}>{item.num_comments}</span>
                <span style={smallColumn}>{item.points}</span>
                <span style={smallColumn}>
                <Button
                    onClick={() => onDismiss(item.objectID)}
                    className="button-inline">
                    Dismiss
                </Button>
            </span>
            </div>
        )}
    </div>;
}

Table.propTypes = {
    list: PropTypes.arrayOf(PropTypes.shape({
        objectID: PropTypes.string.isRequired,
        title: PropTypes.string,
        url: PropTypes.string,
        author: PropTypes.string,
        num_comments: PropTypes.number,
        points: PropTypes.number,
    })).isRequired,
    onDismiss: PropTypes.func
}

const Sort = ({
                  sortKey,
                  activeSortKey,
                  onSort,
                  isSortReverse,
                  children
              }) => {
    const sortClass = classNames(
        'button-inline',
        {'button-active': sortKey === activeSortKey}
    );
    return (
        <div>
            <Button
                onClick={() => onSort(sortKey)}
                className={sortClass}
            >
                {children}
            </Button>
            {
                sortKey === activeSortKey ? (isSortReverse ?
                    <FontAwesomeIcon style={{marginLeft: '10px'}} icon={faSortUp}/> :
                    <FontAwesomeIcon style={{marginLeft: '10px'}} icon={faSortDown}/>) :
                    <FontAwesomeIcon style={{marginLeft: '10px'}} icon={faSort}/>
            }
        </div>
    );
}

export default Table;