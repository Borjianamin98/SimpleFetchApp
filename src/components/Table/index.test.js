import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Table from "./index.js";

describe('Table', () => {
    const props = {
        list: [
            {title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y'},
            {title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z'},
        ]
    };
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Table {...props} />, div);
    });
    it('has a valid snapshot', () => {
        const component = renderer.create(<Table {...props} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});