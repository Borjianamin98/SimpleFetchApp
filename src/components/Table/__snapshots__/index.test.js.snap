// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <div>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Title
        </button>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "marginLeft": "10px",
            }
          }
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <div>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Author
        </button>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "marginLeft": "10px",
            }
          }
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <div>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Comments
        </button>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "marginLeft": "10px",
            }
          }
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <div>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Points
        </button>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "marginLeft": "10px",
            }
          }
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
